core/src/main/scala/org/apache/spark/deploy/master/Master.scala

```diff
@@ -715,7 +715,9 @@ private[deploy] class Master(
         val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
           .filter(canLaunchExecutor(_, app.desc))
           .sortBy(_.coresFree).reverse
-        if (waitingApps.length == 1 && usableWorkers.isEmpty) {
+        val appMayHang = waitingApps.length == 1 &&
```

Comment:
Why do you need to verify `waitingApps.length == 1` ?

Reply:
hmm..I was also thinking about this question when I was fixing. I think we can extend it to multiple apps. Let me try it.

Reply:
I'm not sure how useful is this warning message, this just reflects the current status within one round of scheduling, we can tolerant the behavior to not launch a new executor on currently available workers.

Reply:
It's from https://github.com/apache/spark/pull/25047#discussion_r311779752 originally. The warning is mainly used to detect the possible hang of an application.

Reply:
I don't like this check, but that seems to be the most simple solution for now. Let's keep it then.

Reply:
I think the idea of ==1 is that there is only 1 active application, if more then 1, other applications could be using resources so unless you did a lot more checking. We could definitely make it a more comprehensive check in the future.  I believe this was supposed to be a simple sanity check for people playing around that may have started the cluster with not enough resources.  The standalone mode does not tell you there aren't enough and will just hang forever.